src/segwit_sighashes.rs

```diff
@@ -0,0 +1,236 @@
+// Rust Bitcoin Library
+// Written in 2018 by
+//     Andrew Poelstra <apoelstra@wpsoftware.net>
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! BIP143 Implementation
+//!
+//! Implementation of BIP143 Segwit-style signatures. Should be sufficient
+//! to create signatures for Segwit transactions (which should be pushed into
+//! the appropriate place in the `Transaction::witness` array) or bcash
+//! signatures, which are placed in the scriptSig.
+//!
+
+use encode::{self, Encodable};
+use hash_types::SigHash;
+use hashes::{sha256d, Hash};
+use script::Script;
+use std::ops::Deref;
+use std::io;
+use transaction::SigHashType;
+use transaction::Transaction;
+use confidential;
+
+/// A replacement for SigHashComponents which supports all sighash modes
+pub struct SigHashCache<R: Deref<Target = Transaction>> {
```

Comment:
Given that you are not using any associated types of the trait bound in the field declarations, you could consider removing it from the struct declaration altogether.
All the useful functionality of `SigHashCache` is guarded by the trait bound on the `impl` block anyway.

Having trait bounds on struct declarations can be painful further down the line because they have to be restated everywhere you use the struct, even if you are not using the functionality. It might not matter in this case but it is still unnecessary to include here :)

Reply:
This reasoning makes sense. I blindly adapted the `SigHashCache` code from rust-bitcoin to rust-elements. Maybe @apoelstra knows why have the bound.

Reply:
No, `SigHashCache` was proposed and merged without my knowledge or intervention :)

I'm a little surprised that we don't use the bound anywhere ... how do we compute hashes of transaction data without knowing that we have a transaction? But we can remove it if this is true.

Reply:
> I'm a little surprised that we don't use the bound anywhere ... how do we compute hashes of transaction data without knowing that we have a transaction? But we can remove it if this is true.

The bound is used on the `impl` block!
But you don't need the bound on the `struct` declaration to be able to put a bound on the `impl` block.

That is how most of the stdlib is organized: structs have "plain" type parameters and impl blocks then constrain, which functionality is available given which bounds.

The only exceptions to this are:

- Bounds you need in the `Drop` implementation, those will need to go on the struct as well
- Associated types used to declare more fields as in:

```rust
struct Foo<T: Bar> {
	baz: T::Baz
}
```